runtime/src/main/kotlin/kotlin/text/CodingExceptions.kt

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2010-2017 JetBrains s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kotlin.text
+
+/**
+ *  The exception thrown when a character encoding or decoding error occurs.
+ */
+@SinceKotlin("1.3")
+@ExperimentalStdlibApi
+public actual open class CharacterCodingException(message: String?) : Exception(message) {
+    actual constructor() : this(null)
+}
+
+@UseExperimental(ExperimentalStdlibApi::class)
+public open class IllegalCharacterConversionException : CharacterCodingException {
```

Comment:
The idea is to throw the same exception in both old and new api to avoid code duplication or other kind of problems.
However:
* moving `IllegalCharacterConversionException ` to `kotlin.text` package is a source breaking change
* extending `CharacterCodingException` changes program behaviour if previous superclass (`IllegalArgumentException`) was checked.

Making `CharacterCodingException` to extent `IllegalCharacterConversionException ` may be a better solution.

Reply:
I think we can throw `CharacterCodingException` in the first place, then catch it in the old `...orThrow` functions and replace it with the old `IllegalCharacterConversionException` there.

Reply:
Not sure if compatibility here is too important.

Reply:
The only thing that prevented me from getting completely rid of `IllegalCharacterConversionException` is former api documentation, which explicitly states that an instance of that exception will be thrown.

If you think kotlin-native is currently in a state that such changes do not need backward compatibility, it would be great to avoid deprecation cycle.

Reply:
Yes, there's no need for backward compatibility here.